### PR TITLE
rpcdaemon: fix trace_* API family gas cost

### PR DESCRIPTION
**Pull request by Sixtysixter**
Fix for gas cost in trace API family

### CODE REVIEW

File: silkworm/rpc/core/evm_trace.cpp [hunk 1]
```diff
@@ -576,10 +575,10 @@ void VmTraceTracer::on_execution_start(evmc_revision rev, const evmc_message& ms
...
         metrics_ = evmc_get_instruction_metrics_table(rev);
     }
-    if (precompile::is_precompile(msg.code_address, rev)) {
-        is_precompile_ = true;
-        return;
-    }
+    //    if (precompile::is_precompile(msg.code_address, rev)) {
+    //        is_precompile_ = true;
+    //        return;
+    //    }
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:581]: I suppose this commented code is not necessary anymore, so you should remove it.

File: silkworm/rpc/core/evm_trace.cpp [hunk 2]
```diff
@@ -622,11 +621,24 @@ void VmTraceTracer::on_execution_start(evmc_revision rev, const evmc_message& ms
...
 }
 
-void VmTraceTracer::on_instruction_start(uint32_t pc, const intx::uint256* stack_top, const int /*stack_height*/, const int64_t gas,
-                                         const evmone::ExecutionState& execution_state, const silkworm::IntraBlockState& /*intra_block_state*/) noexcept {
+void VmTraceTracer::on_instruction_start(uint32_t pc, const intx::uint256* stack_top, const int stack_height, const int64_t gas,
+                                         const evmone::ExecutionState& execution_state, const silkworm::IntraBlockState& intra_block_state) noexcept {
     const auto op_code = execution_state.original_code[pc];
     auto op_name = get_opcode_name(opcode_names_, op_code);
 
+    if (fix_call_gas_info_) {  // previuos opcodw was a CALL
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:629]: typo: `opcodw` -> `opcode`

File: silkworm/rpc/core/evm_trace.cpp [hunk 3]
```diff
@@ -747,6 +781,34 @@ void VmTraceTracer::on_pre_check_failed(const evmc_result& /*result*/, const evm
...
+            }
+            SILK_DEBUG << "DebugTracer::evaluate_call_fixes:"
+                       << " call_gas: " << call_gas
+                       << " dst: " << dst
+                       << " value: " << value
+                       << " gas_cost: " << fix_call_gas_info_->gas_cost
+                       << " stipend: " << fix_call_gas_info_->stipend;
+        }
+    }
+}
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.cpp:810]: Some more comments with references to YP or Geth/Erigon implementation would be very helpful here.

File: silkworm/rpc/core/evm_trace.hpp [hunk 4]
```diff
@@ -144,15 +153,21 @@ class VmTraceTracer : public silkworm::EvmTracer {
...
     VmTrace& vm_trace_;
     std::int32_t transaction_index_;
     std::stack<std::string> index_prefix_;
     std::stack<std::reference_wrapper<VmTrace>> traces_stack_;
     const char* const* opcode_names_ = nullptr;
+    const evmc_instruction_metrics* metrics_ = nullptr;
     std::stack<int64_t> start_gas_;
     std::stack<TraceMemory> trace_memory_stack_;
-    const evmc_instruction_metrics* metrics_ = nullptr;
+    std::unique_ptr<FixCallGasInfo> fix_call_gas_info_;
```
Comment (canepat) [inline silkworm/rpc/core/evm_trace.hpp:170]: Let's use `std::optional` instead of `std::unique_ptr` here, as outlined in #2192 for `DebugTracer`.

**Review by canepat — CHANGES_REQUESTED** [review]

File: silkworm/rpc/core/evm_trace.cpp (hunk 1 above)
Comment (Sixtysixter) [reply silkworm/rpc/core/evm_trace.cpp:581]: opps... Correct, removed.

File: silkworm/rpc/core/evm_trace.cpp (hunk 2 above)
Comment (Sixtysixter) [reply silkworm/rpc/core/evm_trace.cpp:629]: Also previuos is wrong.... 
Fixed

File: silkworm/rpc/core/evm_trace.cpp (hunk 3 above)
Comment (Sixtysixter) [reply silkworm/rpc/core/evm_trace.cpp:810]: Maybe better to evmone implementation

File: silkworm/rpc/core/evm_trace.hpp (hunk 4 above)
Comment (Sixtysixter) [reply silkworm/rpc/core/evm_trace.hpp:170]: Done

**Review by canepat — APPROVED** [review]

**Pull request merged**